expanded weights: group norm faster rule

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70141
* __->__ #73696
* #73694

Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

Commit:
expanded weights: group norm faster rule

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"




[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719388).

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719388).

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2294,6 +2294,8 @@
 
 - func: group_norm(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enabled=True) -> Tensor
 
+- func: _group_norm_all_outputs(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05) -> (Tensor, Tensor, Tensor)
```
Yes, we don't expose forward/backward pairs to python.
`torch.nn.grad` was an attempt at that and as you can see from the current state. Never worked out as it is way too tricky.

For this one, why not just move the checks into the `native_group_norm` function (which should be private most likely btw).
They will be duplicated but I feel like this is better than having 3 functions for a single op.

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/group_norm.cpp:
```diff
@@ -31,6 +31,29 @@ std::tuple<Tensor, Tensor, Tensor> native_group_norm(
   const Tensor& gamma = *gamma_maybe_owned;
   const Tensor& beta = c10::value_or_else(beta_opt, [] { return Tensor(); });
 
+  TORCH_CHECK(
+      C % group == 0,
+      "Expected number of channels in input to be divisible by ",
+      "num_groups, but got input of shape ",
+      X.sizes(),
+      " and "
+      "num_groups=",
+      group);
+  TORCH_CHECK(
+      !gamma.defined() || (gamma.dim() == 1 && gamma.numel() == C),
+      "Expected weight to be a vector of size equal to the number of ",
+      "channels in input, but got weight of shape ",
+      gamma.sizes(),
+      " and input of shape ",
+      X.sizes());
+  TORCH_CHECK(
+      !beta.defined() || (beta.dim() == 1 && beta.numel() == C),
+      "Expected bias to be a vector of size equal to the number of ",
+      "channels in input, but got bias of shape ",
+      beta.sizes(),
+      " and input of shape ",
+      X.sizes());
+
```
Is it possible to write a `check_group_norm_inputs(...)` function that does this and share it between native_group_norm and group_norm?

We should also leave a comment about why this shouldn't be deleted (someone will come along and inevitably try to delete the additional check). To defend against deletion we probably want to add a test somewhere for aten.ops.native_group_norm that asserts that it checks for these conditions.

samdow replied:
Moved to helper function, added comment, also added test that expanded weights fails where it should in group norm--I figure that this should be sufficient since we don't exactly care if someone changes what happens in `native_group_norm` if they did it in a way that made expanded weights still correctly fail with incorrect arguments

zou3519 commented on torch/nn/utils/_expanded_weights/group_norm_expanded_weights.py:
```diff
@@ -0,0 +1,58 @@
+        weight, bias, eps = expanded_kwargs['weight'], expanded_kwargs['bias'], expanded_kwargs['eps']
+        output, mean, rstd = forward_helper(torch.native_group_norm, (input, weight, bias, N, C, HxW, num_groups, eps), {})
+        ctx.input, ctx.num_groups = input, num_groups
+        ctx.weight, ctx.bias, ctx.eps = weight, bias, eps
```
Optimization: we don't have to save weight, bias if none of them require gradients, right? (this is the case where someone defines nn.GroupNorm with affine=False.

Doesn't need to be handled in this PR, could be a followup. Also because I am not sure how we'd go about testing this (is there a test case for nn.GroupNorm with affine=False?)

samdow replied:
Added optimization (weight will still be needed for group norm backward, but able to optimize it out if neither input nor weight require gradients)! Group Norm does have an affine=False test but might want to double check when input + weight don't require grad but bias does...

zou3519 reviewed (approved):
I left a suggestion for refactoring the native_group_norm check so that it is shared between native_group_norm and group_norm and adding a test to defend against future people who will try to delete the check. Other than that, this LGTM

Commit:
Update on "expanded weights: group norm faster rule"


Differential Revision: [D34719388](https://our.internmc.facebook.com/intern/diff/D34719388)

[ghstack-poisoned]

samdow commented:
@pytorchbot merge this please
